### PR TITLE
Add body of the request in the cache key

### PR DESCRIPTION
**Pull request by babyba**
edited because on one url we can send requests with different params,
but we get from cache the same value.

i'm not sure about standars, sorry

### CODE REVIEW

**Comment by Kevinrob** [comment]
Thank you @babyba for you PR. I will check it asap!

File: src/Strategy/PrivateCacheStrategy.php [hunk 1]
```diff
@@ -111,8 +111,10 @@ protected function getCacheObject(ResponseInterface $response)
      */
     protected function getCacheKey(RequestInterface $request)
     {
+        $paramsKey = $this->getParamsKey($request);
+
         return sha1(
-            $request->getMethod().$request->getUri()
+            $request->getMethod().$request->getUri().$paramsKey;
```
Comment (Kevinrob) [inline src/Strategy/PrivateCacheStrategy.php]: @babyba you have to remove this `;` here :wink:

Comment (babyba) [inline src/Strategy/PrivateCacheStrategy.php]: thanks, i saw :) sorry, i hadn't enough time ;)

**Comment by Kevinrob** [comment]
@babyba 
The RFC:
>The primary cache key consists of the request method and target URI.

By the way, by default we only store `GET` request and the server must not change the responses based on the body of a `GET` request.

Why would you like to add this change? The URI part already contains the query parameters.

**Comment by babyba** [comment]
Thanks for explanation. I added it because the I had situation when I had to send POST requests with diffrent bodies, but server gave different replies depend on body of POST request, but I guess, it's talk now more about my needs, than functionality and requitments of RFC.

**Comment by Kevinrob** [comment]
@babyba 
Okay, but if you make a POST request to a server, the server have to received the request. The method POST is used to make a action on the server (add content, modify, etc.).
If you cache the response of a POST request, the action will not be executed the next time. For action that will just retrieve data without any modification, the GET method is the right method.

Of course, I know nothing about your situation. Maybe you don't control the server.
If you would still like to cache the response to a POST request, you have to:
- adding `POST` to the http methods in `CacheMiddleware` with `setHttpMethods(['GET' => true, 'POST' => true]);`
- extending `PrivateCacheStrategy` and override `protected function getCacheKey(RequestInterface $request)` for taking the body in the hash

I will close this PR because it's not correct with the RFC. Thank you for your interest!

**Comment by jeronimo93** [comment]
Should the documentation mention that to cache POST requests you have explicitly set this? 

It would be a big help, as currently there isn't anything mentioned and it is not shown in the examples. Took me a little while to work out why my POST requests were not getting cached. Could be me being a n00b, but could have saved me a lot of time if this was documented.

**Comment by Kevinrob** [comment]
@jeromegamez I don't think so. Caching POST is a very bad practice. POST existing for posting something. The request must go to the server.

**Comment by jeronimo93** [comment]
@Kevinrob very true! But I am working with what feels like the worst designed API in the world, which requires making POST requests to 'get' data *facepalm*.

**Comment by babyba** [comment]
@jeronimo93 buy the you can took this functionality to cache post request, i had to work with kind of API and there i couldn't give all parameters into get request, GET size were too big, so I wrote this feature for it and in project which i did, still this functionality.

But @Kevinrob is right, caching post is bad idea. also as you could read above this discussion - we should cache post request in according with standards

**Pull request closed without merging**